File: testing-framework/backend/src/routes/attachments.ts
import { Router } from "express";
import { prisma } from "../prisma.js";
import path from "path";
import { existsSync, mkdirSync, writeFileSync } from "fs";
import multer from "multer";
import { randomUUID } from "crypto";
import { parseAttachment } from "../services/attachmentParser.js";

export const attachmentsRouter = Router();

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 20 * 1024 * 1024 },
});

attachmentsRouter.post("/upload", upload.array("files", 20), async (req, res) => {
  const requirementId = req.body?.requirementId as string;
  const files = (req as unknown as { files: Express.Multer.File[] }).files;
  if (!requirementId) return res.status(400).json({ error: "requirementId is required" });
  if (!files?.length) return res.status(400).json({ error: "至少上传一个文件" });
  const requirement = await prisma.requirement.findUnique({ where: { id: requirementId } });
  if (!requirement) return res.status(404).json({ error: "需求不存在" });
  const uploadsDir = path.join(process.cwd(), "uploads", requirementId);
  if (!existsSync(uploadsDir)) mkdirSync(uploadsDir, { recursive: true });
  const created: { id: string; filename: string }[] = [];
  for (const file of files) {
    const originalname =
      Buffer.from(file.originalname, "latin1").toString("utf8") || file.originalname;
    const ext = path.extname(originalname) || "";
    const base = path.basename(originalname, ext) || "file";
    const safeName = `${base}-${randomUUID().slice(0, 8)}${ext}`;
    const filePath = path.join(uploadsDir, safeName);
    writeFileSync(filePath, file.buffer);
    const a = await prisma.requirementAttachment.create({
      data: {
        requirementId,
        filename: originalname,
        filePath: path.relative(process.cwd(), filePath),
        mimeType: file.mimetype || "",
        size: file.size,
      },
    });
    const extracted = await parseAttachment(file.buffer, file.mimetype || "", originalname);
    if (extracted != null) {
      await prisma.requirementAttachment.update({
        where: { id: a.id },
        data: { extractedText: extracted },
      });
    } else {
      console.warn(`[上传附件] "${originalname}" 解析结果为空 (mimeType=${file.mimetype || ""})`);
    }
    created.push({ id: a.id, filename: a.filename });
  }
  res.json({ created });
});

attachmentsRouter.get("/:id/file", async (req, res) => {
  const a = await prisma.requirementAttachment.findUnique({
    where: { id: req.params.id },
  });
  if (!a) return res.status(404).send();
  const fullPath = path.resolve(a.filePath);
  if (!existsSync(fullPath)) return res.status(404).send();
  if (req.query.download === "1") {
    const encoded = encodeURIComponent(a.filename);
    res.setHeader(
      "Content-Disposition",
      `attachment; filename*=UTF-8''${encoded}`
    );
  }
  res.sendFile(fullPath);
});

File: testing-framework/backend_python/app/routers/attachments.py
from __future__ import annotations

import uuid
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session

from app.config import BACKEND_CWD
from app.database import get_db
from app.models import Requirement, RequirementAttachment
from app.services.attachment_parser import parse_attachment
from app.util import new_id

router = APIRouter(prefix="/attachments", tags=["attachments"])


@router.post("/upload")
async def upload_attachments(
    requirementId: str = Form(...),
    files: list[UploadFile] = File(...),
    db: Session = Depends(get_db),
):
    if not requirementId:
        raise HTTPException(status_code=400, detail="requirementId is required")
    if not files:
        raise HTTPException(status_code=400, detail="至少上传一个文件")

    req = db.query(Requirement).filter(Requirement.id == requirementId).first()
    if not req:
        raise HTTPException(status_code=404, detail="需求不存在")

    uploads_dir = Path(BACKEND_CWD) / "uploads" / requirementId
    uploads_dir.mkdir(parents=True, exist_ok=True)
    created: list[dict[str, str]] = []

    for file in files:
        originalname = file.filename or "file"
        ext = Path(originalname).suffix or ""
        base = Path(originalname).stem or "file"
        safe_name = f"{base}-{uuid.uuid4().hex[:8]}{ext}"
        dest = uploads_dir / safe_name
        data = await file.read()
        dest.write_bytes(data)
        rel = str(dest.relative_to(Path(BACKEND_CWD)))

        att = RequirementAttachment(
            id=new_id(),
            requirementId=requirementId,
            filename=originalname,
            filePath=rel,
            mimeType=file.content_type or "",
            size=len(data),
        )
        db.add(att)
        db.flush()

        extracted = parse_attachment(data, file.content_type or "", originalname)
        if extracted is not None:
            att.extractedText = extracted
        else:
            print(f'[上传附件] "{originalname}" 解析结果为空 (mimeType={file.content_type or ""})')

        created.append({"id": att.id, "filename": att.filename})

    db.commit()
    return {"created": created}


@router.get("/{att_id}/file")
def get_attachment_file(att_id: str, download: Optional[str] = None, db: Session = Depends(get_db)):
    a = db.query(RequirementAttachment).filter(RequirementAttachment.id == att_id).first()
    if not a:
        raise HTTPException(status_code=404, detail="Not found")
    full_path = (Path(BACKEND_CWD) / a.filePath).resolve()
    try:
        full_path.relative_to(Path(BACKEND_CWD).resolve())
    except ValueError:
        raise HTTPException(status_code=404, detail="Not found")
    if not full_path.is_file():
        raise HTTPException(status_code=404, detail="Not found")

    headers = {}
    if download == "1":
        from urllib.parse import quote

        enc = quote(a.filename)
        headers["Content-Disposition"] = f"attachment; filename*=UTF-8''{enc}"
    return FileResponse(str(full_path), filename=a.filename, headers=headers)
